# Task

## T5.1: Download the 'Portugal_online_retail', 'Sweden_online_retail, and 'UK_online_retail' datasets from Canvas. Apply the apriori algorithm to all datasets using three different confidence levels. Select one confidence level for each dataset that you think works better. Determine the first three most important rules for each dataset using the selected confidence level and report them in the report cell. Explain what each rule means (Completing the report cell is required) (15%).

NOTE: You should comment on your code

In [15]:
############### Write your code in this cell (If applicable) ##################
# Importing the relevant libraries
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

# Reading csv file into a pandas DataFrame
basket_Portugal = pd.read_csv('C:/Users/HP/Downloads/Portugal_online_retail.csv')

#printing data
basket_Portugal

InvoiceNo  10 COLOUR SPACEBOY PEN  12 PENCIL SMALL TUBE WOODLAND  \
0      536990                       0                              0   
1      537246                       0                              0   
2      537818                       0                              0   
3      537915                       0                              0   
4      538311                       0                              0   
5      539353                       0                              0   
6      540519                       0                              0   
7      540546                       1                              0   
8      541430                       0                              0   
9      542147                       0                              0   
10     544495                       0                              0   
11     545191                       0                              0   
12     545937                       0                              0   
13     547444                       0                              0   
14     547566                       0                              0   
15     547897                       0                              0   
16     548470                       0                              0   
17     549435                       0                              0   
18     549953                       0                              0   
19     553017                       0                              1   
20     553018                       0                              0   
21     554155                       0                              0   
22     554301                       0                              0   
23     554305                       0                              0   
24     555862                       0                              0   
25     556024                       0                              0   
26     556829                       0                              0   
27     559862                       0                              0   
28     559907                       0                              0   
29     560694                       0                              0   
30     563241                       0                              0   
31     565295                       0                              0   
32     567929                       0                              0   
33     568208                       0                              0   
34     569866                       0                              1   
35     569870                       0                              0   
36     570499                       0                              0   
37     570964                       0                              0   
38     571188                       0                              0   
39     571324                       0                              0   
40     571326                       0                              0   
41     571477                       0                              0   
42     571901                       0                              0   
43     572680                       0                              0   
44     574099                       0                              0   
45     574844                       0                              0   
46     575743                       0                              0   
47     576257                       0                              0   
48     576269                       0                              0   
49     576895                       0                              0   
50     577618                       0                              0   
51     578463                       0                              0   
52     578488                       0                              0   
53     579790                       0                              0   
54     580153                       0       

In [16]:
#dropping the Invoice No Column
basket_Portugal=basket_Portugal.drop('InvoiceNo',axis=1)

#printing data to check that invoice no is dropped
basket_Portugal

10 COLOUR SPACEBOY PEN  12 PENCIL SMALL TUBE WOODLAND  \
0                        0                              0   
1                        0                              0   
2                        0                              0   
3                        0                              0   
4                        0                              0   
5                        0                              0   
6                        0                              0   
7                        1                              0   
8                        0                              0   
9                        0                              0   
10                       0                              0   
11                       0                              0   
12                       0                              0   
13                       0                              0   
14                       0                              0   
15                       0                              0   
16                       0                              0   
17                       0                              0   
18                       0                              0   
19                       0                              1   
20                       0                              0   
21                       0                              0   
22                       0                              0   
23                       0                              0   
24                       0                              0   
25                       0                              0   
26                       0                              0   
27                       0                              0   
28                       0                              0   
29                       0                              0   
30                       0                              0   
31                       0                              0   
32                       0                              0   
33                       0                              0   
34                       0                              1   
35                       0                              0   
36                       0                              0   
37                       0                              0   
38                       0                              0   
39                       0                              0   
40                       0                              0   
41                       0                              0   
42                       0                              0   
43                       0                              0   
44                       0                              0   
45                       0                              0   
46                       0                              0   
47                       0                              0   
48                       0                              0   
49                       0                              0   
50                       0                              0   
51                       0                              0   
52                       0                              0   
53                       0                              0   
54                       0                              0   
55                       0                              0   
56                       0                              0   
57                       0                              0   

    12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE POSY  \
0                                     0                          0   
1                                     0                          0   
2                                     0                          0   
3                                     0                          0   
4                                     0               

In [17]:
# Finding the frequent itemsets
frq_items = apriori(basket_Portugal, min_support = 0.15, use_colnames = True)

#Sorting and printing the values in descending order
frq_items.sort_values(by=['support'], ascending = False)

C:\Users\HP\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
8   0.517241                                          (POSTAGE)
6   0.241379                          (LUNCH BAG RED RETROSPOT)
10  0.241379                  (RETROSPOT TEA SET CERAMIC 11 PC)
0   0.206897                     (BAKING SET 9 PIECE RETROSPOT)
5   0.206897                              (LUNCH BAG CARS BLUE)
4   0.189655                (JUMBO SHOPPER VINTAGE RED PAISLEY)
1   0.172414                        (CHARLOTTE BAG SUKI DESIGN)
7   0.172414                  (PLASTERS IN TIN VINTAGE PAISLEY)
9   0.172414                      (RED RETROSPOT CHARLOTTE BAG)
2   0.155172                   (JUMBO BAG PINK VINTAGE PAISLEY)
3   0.155172              (JUMBO BAG SCANDINAVIAN BLUE PAISLEY)
11  0.155172  (RETROSPOT TEA SET CERAMIC 11 PC, BAKING SET 9...
12  0.155172  (JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG ...

In [18]:
#printing shape 
frq_items.shape

(13, 2)

In [19]:
# Collecting the inferred rules and defining a confidence level of 60%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.6)

#sorting and printing values in descending order of confidence level
rules.sort_values(by=['confidence'], ascending = False)

antecedents                          consequents  \
3     (JUMBO BAG PINK VINTAGE PAISLEY)  (JUMBO SHOPPER VINTAGE RED PAISLEY)   
2  (JUMBO SHOPPER VINTAGE RED PAISLEY)     (JUMBO BAG PINK VINTAGE PAISLEY)   
1       (BAKING SET 9 PIECE RETROSPOT)    (RETROSPOT TEA SET CERAMIC 11 PC)   
0    (RETROSPOT TEA SET CERAMIC 11 PC)       (BAKING SET 9 PIECE RETROSPOT)   

   antecedent support  consequent support   support  confidence      lift  \
3            0.155172            0.189655  0.155172    1.000000  5.272727   
2            0.189655            0.155172  0.155172    0.818182  5.272727   
1            0.206897            0.241379  0.155172    0.750000  3.107143   
0            0.241379            0.206897  0.155172    0.642857  3.107143   

   leverage  conviction  zhangs_metric  
3  0.125743         inf       0.959184  
2  0.125743    4.646552       1.000000  
1  0.105232    3.034483       0.855072  
0  0.105232    2.220690       0.893939

In [20]:
# Collecting the inferred rules and defining a confidence level of 70%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.7)

#Sorting and printing rules by confidence in descending order 
rules.sort_values(by=['confidence'], ascending = False)

antecedents                          consequents  \
2     (JUMBO BAG PINK VINTAGE PAISLEY)  (JUMBO SHOPPER VINTAGE RED PAISLEY)   
1  (JUMBO SHOPPER VINTAGE RED PAISLEY)     (JUMBO BAG PINK VINTAGE PAISLEY)   
0       (BAKING SET 9 PIECE RETROSPOT)    (RETROSPOT TEA SET CERAMIC 11 PC)   

   antecedent support  consequent support   support  confidence      lift  \
2            0.155172            0.189655  0.155172    1.000000  5.272727   
1            0.189655            0.155172  0.155172    0.818182  5.272727   
0            0.206897            0.241379  0.155172    0.750000  3.107143   

   leverage  conviction  zhangs_metric  
2  0.125743         inf       0.959184  
1  0.125743    4.646552       1.000000  
0  0.105232    3.034483       0.855072

In [21]:
# Collecting the inferred rules and defining a confidence level of 80%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.8)

#Sorting and printing rules by confidence in descending order 
rules.sort_values(by=['confidence'], ascending = False)

antecedents                          consequents  \
1     (JUMBO BAG PINK VINTAGE PAISLEY)  (JUMBO SHOPPER VINTAGE RED PAISLEY)   
0  (JUMBO SHOPPER VINTAGE RED PAISLEY)     (JUMBO BAG PINK VINTAGE PAISLEY)   

   antecedent support  consequent support   support  confidence      lift  \
1            0.155172            0.189655  0.155172    1.000000  5.272727   
0            0.189655            0.155172  0.155172    0.818182  5.272727   

   leverage  conviction  zhangs_metric  
1  0.125743         inf       0.959184  
0  0.125743    4.646552       1.000000

######################## Write your report in this cell (if applicable) #############################

For the Portugal dataset, i imported the relevant libraries which includes numpy, pandas, apriori and association rules. Then went further to read my csv file and preprocessed the data by dropping the invoice_No column. To extract the frequent item sets in the basket_Portugal dataset, i set the min_support level to 0.15 for an efficient output of 13 rows, after some adjustment from 0.1 which gave an output of 110 rows and 0.12 which gave an output of 55 rows. Consequently, all item sets with a support level below 0.15 were not showcased. To collect the inferred rules, with the association code, i set the measuring metric to confidence and selected three (3) confidence levels of 0.6, 0.7 and 0.8 simultaneously. Here, 0.6 as minimum threshold for confidence level is the best as it gave more options to choose the best 3 item sets with strong association.

RULE 1.
The first most important 2 items association rule result with 0.6 confidence shows that if customers buy JUMBO BAG PINK VINTAGE PAISLEY, they are 100% highly likely to buy JUMBO SHOPPER VINTAGE RED PAISLEY.

RULE 2
The second most important 2 items association rule result shows that if customers buy JUMBO SHOPPER VINTAGE RED PAISLEY, they are 81% likely to buy JUMBO BAG PINK VINTAGE PAISLEY

RULE 3
The Third most important 2 items association rule shows that if customers buy BAKING SET 9 PIECE RETROSPOT, they are 75% likely to buy RETROSPOT TEA SET CERAMIC 11 PC (246 words)

In [22]:
# Reading Basket_Sweden csv file
basket_Sweden = pd.read_csv('C:/Users/HP/Downloads/Sweden_online_retail.csv')

#printing data
basket_Sweden

InvoiceNo  12 PENCIL SMALL TUBE WOODLAND  \
0      538848                              0   
1      539338                              0   
2      540040                              0   
3      542428                              0   
4      542911                              0   
5      546161                              0   
6      546530                              0   
7      547645                              0   
8      550277                              0   
9      551278                              0   
10     552717                              0   
11     552957                              0   
12     553674                              0   
13     555403                              0   
14     556194                              0   
15     557939                              0   
16     559145                              0   
17     561890                              0   
18     561902                              0   
19     562285                              0   
20     562600                              0   
21     562650                              0   
22     563443                              0   
23     564539                              0   
24     565287                              0   
25     566494                              0   
26     569018                              0   
27     571191                              0   
28     571318                              0   
29     573333                              0   
30     574690                              1   
31     575736                              0   
32     576083                              0   
33     576519                              0   
34     578089                              0   
35     580704                              0   

    12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
0                                     0                            0   
1                                     0                            0   
2                                     0                            0   
3                                     0                            0   
4                                     0                            0   
5                                     0                            0   
6                                     1                            1   
7                                     0                            0   
8                                     0                            0   
9                                     0                            0   
10                                    0                            0   
11                                    0                            0   
12                                    0                            0   
13                                    0                            0   
14                                    0                            0   
15                                    0                            0   
16                                    0                            0   
17                                    0                            0   
18                                    0                            0   
19                                    0                            0   
20                                    0                            1   
21                                    0                            0   
22                                    0                            0   
23                                    0                            0   
24                                    0                            0   
25                                    0                            0   
26                                    0                            0   
27                                    0                            0   
28                                    0                            0   
29                                    0                            

In [23]:
#dropping the Invoice No Column
basket_Sweden = basket_Sweden.drop('InvoiceNo',axis=1)

#printing data to check that its dropped
basket_Sweden

12 PENCIL SMALL TUBE WOODLAND  12 PENCILS SMALL TUBE RED RETROSPOT  \
0                               0                                    0   
1                               0                                    0   
2                               0                                    0   
3                               0                                    0   
4                               0                                    0   
5                               0                                    0   
6                               0                                    1   
7                               0                                    0   
8                               0                                    0   
9                               0                                    0   
10                              0                                    0   
11                              0                                    0   
12                              0                                    0   
13                              0                                    0   
14                              0                                    0   
15                              0                                    0   
16                              0                                    0   
17                              0                                    0   
18                              0                                    0   
19                              0                                    0   
20                              0                                    0   
21                              0                                    0   
22                              0                                    0   
23                              0                                    0   
24                              0                                    0   
25                              0                                    0   
26                              0                                    0   
27                              0                                    0   
28                              0                                    0   
29                              0                                    0   
30                              1                                    0   
31                              0                                    0   
32                              0                                    0   
33                              0                                    0   
34                              0                                    0   
35                              0                                    0   

    12 PENCILS SMALL TUBE SKULL  12 PENCILS TALL TUBE WOODLAND  \
0                             0                              0   
1                             0                              0   
2                             0                              0   
3                             0                              0   
4                             0                              0   
5                             0                              0   
6                             1                              0   
7                             0                              0   
8                             0                              0   
9                             0                              0   
10                            0                              0   
11                            0                              0   
12                            0                              0   
13                            0                              0   
14                            0                              0   
15                            0                              0   
16                            0                              0   
17                            0                              0   
18         

In [24]:
# Finding the frequent itemsets
frq_items = apriori(basket_Sweden, min_support = 0.1, use_colnames = True)

#Sorting the values in descending order
frq_items.sort_values(by=['support'], ascending = False)

C:\Users\HP\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
8   0.611111                                          (POSTAGE)
14  0.194444                 (SET OF 3 CAKE TINS PANTRY DESIGN)
4   0.166667                           (MINI PAINT SET VINTAGE)
2   0.138889                                (GUMBALL COAT RACK)
3   0.138889                           (JAM MAKING SET PRINTED)
23  0.138889           (POSTAGE, RED TOADSTOOL LED NIGHT LIGHT)
20  0.138889                       (POSTAGE, GUMBALL COAT RACK)
10  0.138889                    (RED TOADSTOOL LED NIGHT LIGHT)
19  0.138889                (WORLD WAR 2 GLIDERS ASSTD DESIGNS)
0   0.111111                  (60 CAKE CASES DOLLY GIRL DESIGN)
16  0.111111                              (WALL TIDY RETROSPOT)
22  0.111111                      (POSTAGE, RABBIT NIGHT LIGHT)
21  0.111111  (PACK OF 72 RETROSPOT CAKE CASES, PACK OF 60 S...
18  0.111111                        (WOODEN OWLS LIGHT GARLAND)
17  0.111111                           (WOODEN BOX OF DOMINOES)
12  0.111111                (ROUND SNACK BOXES SET OF 4 FRUITS)
15  0.111111               (SET OF 72 RETROSPOT PAPER  DOILIES)
13  0.111111                (SET OF 12 FAIRY CAKE BAKING CASES)
1   0.111111                  (60 CAKE CASES VINTAGE CHRISTMAS)
11  0.111111                  (RETROSPOT TEA SET CERAMIC 11 PC)
9   0.111111                               (RABBIT NIGHT LIGHT)
7   0.111111                  (PACK OF 72 RETROSPOT CAKE CASES)
6   0.111111                   (PACK OF 60 SPACEBOY CAKE CASES)
5   0.111111               (PACK OF 60 PINK PAISLEY CAKE CASES)
24  0.111111                     (WALL TIDY RETROSPOT, POSTAGE)

In [25]:
#printing the shape to check efficiency
frq_items.shape

(25, 2)

In [26]:
# Collecting the inferred rules and defining a confidence level of 60%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.6)

#Sorting values by descending order 
rules.sort_values(by=['confidence'], ascending = False)

antecedents                        consequents  \
0                (GUMBALL COAT RACK)                          (POSTAGE)   
1  (PACK OF 72 RETROSPOT CAKE CASES)   (PACK OF 60 SPACEBOY CAKE CASES)   
2   (PACK OF 60 SPACEBOY CAKE CASES)  (PACK OF 72 RETROSPOT CAKE CASES)   
3               (RABBIT NIGHT LIGHT)                          (POSTAGE)   
4    (RED TOADSTOOL LED NIGHT LIGHT)                          (POSTAGE)   
5              (WALL TIDY RETROSPOT)                          (POSTAGE)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.138889            0.611111  0.138889         1.0  1.636364   
1            0.111111            0.111111  0.111111         1.0  9.000000   
2            0.111111            0.111111  0.111111         1.0  9.000000   
3            0.111111            0.611111  0.111111         1.0  1.636364   
4            0.138889            0.611111  0.138889         1.0  1.636364   
5            0.111111            0.611111  0.111111         1.0  1.636364   

   leverage  conviction  zhangs_metric  
0  0.054012         inf       0.451613  
1  0.098765         inf       1.000000  
2  0.098765         inf       1.000000  
3  0.043210         inf       0.437500  
4  0.054012         inf       0.451613  
5  0.043210         inf       0.437500

In [27]:
# Collecting the inferred rules and defining a confidence level of 70%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.7)

#Sorting values by descending order of highest to lowest
rules.sort_values(by=['confidence'], ascending = False)

antecedents                        consequents  \
0                (GUMBALL COAT RACK)                          (POSTAGE)   
1  (PACK OF 72 RETROSPOT CAKE CASES)   (PACK OF 60 SPACEBOY CAKE CASES)   
2   (PACK OF 60 SPACEBOY CAKE CASES)  (PACK OF 72 RETROSPOT CAKE CASES)   
3               (RABBIT NIGHT LIGHT)                          (POSTAGE)   
4    (RED TOADSTOOL LED NIGHT LIGHT)                          (POSTAGE)   
5              (WALL TIDY RETROSPOT)                          (POSTAGE)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.138889            0.611111  0.138889         1.0  1.636364   
1            0.111111            0.111111  0.111111         1.0  9.000000   
2            0.111111            0.111111  0.111111         1.0  9.000000   
3            0.111111            0.611111  0.111111         1.0  1.636364   
4            0.138889            0.611111  0.138889         1.0  1.636364   
5            0.111111            0.611111  0.111111         1.0  1.636364   

   leverage  conviction  zhangs_metric  
0  0.054012         inf       0.451613  
1  0.098765         inf       1.000000  
2  0.098765         inf       1.000000  
3  0.043210         inf       0.437500  
4  0.054012         inf       0.451613  
5  0.043210         inf       0.437500

In [28]:
# Collecting the inferred rules and defining a confidence level of 80%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.8)

#Sorting values by descending order of highest to lowest
rules.sort_values(by=['confidence'], ascending = False)

antecedents                        consequents  \
0                (GUMBALL COAT RACK)                          (POSTAGE)   
1  (PACK OF 72 RETROSPOT CAKE CASES)   (PACK OF 60 SPACEBOY CAKE CASES)   
2   (PACK OF 60 SPACEBOY CAKE CASES)  (PACK OF 72 RETROSPOT CAKE CASES)   
3               (RABBIT NIGHT LIGHT)                          (POSTAGE)   
4    (RED TOADSTOOL LED NIGHT LIGHT)                          (POSTAGE)   
5              (WALL TIDY RETROSPOT)                          (POSTAGE)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.138889            0.611111  0.138889         1.0  1.636364   
1            0.111111            0.111111  0.111111         1.0  9.000000   
2            0.111111            0.111111  0.111111         1.0  9.000000   
3            0.111111            0.611111  0.111111         1.0  1.636364   
4            0.138889            0.611111  0.138889         1.0  1.636364   
5            0.111111            0.611111  0.111111         1.0  1.636364   

   leverage  conviction  zhangs_metric  
0  0.054012         inf       0.451613  
1  0.098765         inf       1.000000  
2  0.098765         inf       1.000000  
3  0.043210         inf       0.437500  
4  0.054012         inf       0.451613  
5  0.043210         inf       0.437500

#################################################        REPORT      #########################################################
For the Basket_Sweden dataset, i read the csv file and preprocessed the data by dropping the invoice_No column. To extract the frequent item sets in the basket_Sweden dataset, i set the min_support level to 0.1 for an efficient output of 25 rows, after some adjustment from 0.15 which gave an output of 3 rows and showed no item set association rule. min_support of 0.13 gave an output of 9 rows and showed only 2 rows of association rule which was not also efficient. Consequently, all item sets with a support level below 0.1 were not showcased. To collect the inferred rules, with the association rule code, i set the measuring metric to confidence and selected three (3) confidence levels of 0.6, 0.7 and 0.8 simultaneously. Here, all the 3 confidence levels selected produced the same first 3 output and number of rules with strong association and support.

Rule 1
The first most important 2 items association rule result with all the 3 confidence levels shows that if customers buy GUMBALL COAT RACK, they are 100% highly likely to buy POSTAGE. This occurs 13.89% of the time

Rule 2
The second most important 2 items association rule result with all the 3 confidence levels shows that if customers buy PACK OF 72 RETROSPOTT CAKE CASES, they are 100% highly likely to buy PACK OF 60 SPACEBOY CAKE CASES. This happens 11% of the time

Rule 3
The Third most important 2 items association rule shows that if customers buy PACK OF 60 SPACEBOY CAKE CASES, they are also 100% highly likely to buy PACK OF 72 RETROSPOT CAKE CASES (268 words)

In [29]:
# Reading Basket_UK csv file
basket_UK = pd.read_csv('C:/Users/HP/Downloads/UK_online_retail.csv')

basket_UK

C:\Users\HP\AppData\Local\Temp\ipykernel_6556\4077118609.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  basket_UK = pd.read_csv('C:/Users/HP/Downloads/UK_online_retail.csv')


InvoiceNo  *Boombox Ipod Classic  *USB Office Mirror Ball  \
0        536365                      0                        0   
1        536366                      0                        0   
2        536367                      0                        0   
3        536368                      0                        0   
4        536369                      0                        0   
...         ...                    ...                      ...   
18662    581585                      0                        0   
18663    581586                      0                        0   
18664   A563185                      0                        0   
18665   A563186                      0                        0   
18666   A563187                      0                        0   

       10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   
...                       ...                         ...   
18662                       0                           0   
18663                       0                           0   
18664                       0                           0   
18665                       0                           0   
18666                       0                           0   

       12 DAISY PEGS IN WOOD BOX  12 EGG HOUSE PAINTED WOOD  \
0                              0                          0   
1                              0                          0   
2                              0                          0   
3                              0                          0   
4                              0                          0   
...                          ...                        ...   
18662                          0                          0   
18663                          0                          0   
18664                          0                          0   
18665                          0                          0   
18666                          0                          0   

       12 HANGING EGGS HAND PAINTED  12 IVORY ROSE PEG PLACE SETTINGS  \
0                                 0                                 0   
1                                 0                                 0   
2                                 0                                 0   
3                                 0                                 0   
4                                 0                                 0   
...                             ...                               ...   
18662                             0                                 0   
18663                             0                                 0   
18664                             0                                 0   
18665                             0                                 0   
18666                             0                                 0   

       12 MESSAGE CARDS WITH ENVELOPES  ...  wrongly coded 20713  \
0                                    0  ...                    0   
1                                    0  ...                    0   
2                                    0  ...                    0   
3                                    0  ...                    0   
4                                    0  ...                    0   
...                                ...  ...                  ...   
18662                                0  ...                    0   
18663                                0  ...                    0   
18664                                0  ...                    0   
18665                                0  ...                    0   
18666                                0  ...                    0   

       wrongly co

In [30]:
#dropping the Invoice No Column
basket_UK = basket_UK.drop('InvoiceNo',axis=1)

#Printing data to check that invoice no is dropped
basket_UK

*Boombox Ipod Classic  *USB Office Mirror Ball  10 COLOUR SPACEBOY PEN  \
0                          0                        0                       0   
1                          0                        0                       0   
2                          0                        0                       0   
3                          0                        0                       0   
4                          0                        0                       0   
...                      ...                      ...                     ...   
18662                      0                        0                       0   
18663                      0                        0                       0   
18664                      0                        0                       0   
18665                      0                        0                       0   
18666                      0                        0                       0   

       12 COLOURED PARTY BALLOONS  12 DAISY PEGS IN WOOD BOX  \
0                               0                          0   
1                               0                          0   
2                               0                          0   
3                               0                          0   
4                               0                          0   
...                           ...                        ...   
18662                           0                          0   
18663                           0                          0   
18664                           0                          0   
18665                           0                          0   
18666                           0                          0   

       12 EGG HOUSE PAINTED WOOD  12 HANGING EGGS HAND PAINTED  \
0                              0                             0   
1                              0                             0   
2                              0                             0   
3                              0                             0   
4                              0                             0   
...                          ...                           ...   
18662                          0                             0   
18663                          0                             0   
18664                          0                             0   
18665                          0                             0   
18666                          0                             0   

       12 IVORY ROSE PEG PLACE SETTINGS  12 MESSAGE CARDS WITH ENVELOPES  \
0                                     0                                0   
1                                     0                                0   
2                                     0                                0   
3                                     0                                0   
4                                     0                                0   
...                                 ...                              ...   
18662                                 0                                0   
18663                                 0                                0   
18664                                 0                                0   
18665                                 0                                0   
18666                                 0                                0   

       12 PENCIL SMALL TUBE WOODLAND  ...  wrongly coded 20713  \
0                                  0  ...                    0   
1                                  0  ...                    0   
2                                  0  ...                    0   
3                                  0  ...                    0   
4                                  0  ...                    0   
...                              ...  ...                  ...   
18662                              0  ...                    0   
18663                          

In [31]:
# Finding the frequent itemsets
frq_items = apriori(basket_UK, min_support = 0.03, use_colnames = True)

#Sorting the values in descending order
frq_items.sort_values(by=['support'], ascending = False)

C:\Users\HP\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
118  0.116034               (WHITE HANGING HEART T-LIGHT HOLDER)
52   0.103820                          (JUMBO BAG RED RETROSPOT)
95   0.090266                         (REGENCY CAKESTAND 3 TIER)
84   0.085391                                    (PARTY BUNTING)
69   0.074570                          (LUNCH BAG RED RETROSPOT)
..        ...                                                ...
126  0.030535  (JUMBO  BAG BAROQUE BLACK WHITE, JUMBO BAG RED...
96   0.030428                 (RETROSPOT HEART HOT WATER BOTTLE)
18   0.030214                                   (DOORMAT HEARTS)
123  0.030160  (ALARM CLOCK BAKELIKE RED, ALARM CLOCK BAKELIK...
20   0.030107                              (DOORMAT NEW ENGLAND)

[131 rows x 2 columns]

In [32]:
#Printing data shape
frq_items.shape

(131, 2)

In [33]:
# Collecting the inferred rules and defining a confidence level of 60%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.6)

#Sorting values by descending order of highest to lowest
rules.sort_values(by=['confidence'], ascending = False)

antecedents                        consequents  \
2   (PINK REGENCY TEACUP AND SAUCER)  (GREEN REGENCY TEACUP AND SAUCER)   
4  (GREEN REGENCY TEACUP AND SAUCER)  (ROSES REGENCY TEACUP AND SAUCER)   
5  (ROSES REGENCY TEACUP AND SAUCER)  (GREEN REGENCY TEACUP AND SAUCER)   
7          (JUMBO BAG PINK POLKADOT)          (JUMBO BAG RED RETROSPOT)   
1       (ALARM CLOCK BAKELIKE GREEN)         (ALARM CLOCK BAKELIKE RED)   
6   (JUMBO  BAG BAROQUE BLACK WHITE)          (JUMBO BAG RED RETROSPOT)   
3  (GREEN REGENCY TEACUP AND SAUCER)   (PINK REGENCY TEACUP AND SAUCER)   
8           (JUMBO STORAGE BAG SUKI)          (JUMBO BAG RED RETROSPOT)   
0         (ALARM CLOCK BAKELIKE RED)       (ALARM CLOCK BAKELIKE GREEN)   

   antecedent support  consequent support   support  confidence       lift  \
2            0.037660            0.050035  0.030910    0.820768  16.403939   
4            0.050035            0.051267  0.037553    0.750535  14.639752   
5            0.051267            0.050035  0.037553    0.732497  14.639752   
7            0.062088            0.103820  0.042053    0.677308   6.523895   
1            0.046928            0.049821  0.030160    0.642694  12.900183   
6            0.048749            0.103820  0.030535    0.626374   6.033290   
3            0.050035            0.037660  0.030910    0.617773  16.403939   
8            0.060535            0.103820  0.037392    0.617699   5.949737   
0            0.049821            0.046928  0.030160    0.605376  12.900183   

   leverage  conviction  zhangs_metric  
2  0.029026    5.300203       0.975787  
4  0.034988    3.803076       0.980765  
5  0.034988    3.551237       0.982039  
7  0.035607    2.777201       0.902769  
1  0.027822    2.659288       0.967903  
6  0.025474    2.398601       0.877006  
3  0.029026    2.517719       0.988498  
8  0.031108    2.344176       0.885531  
0  0.027822    2.415142       0.970850

In [34]:
# Collecting the inferred rules and defining a confidence level of 70%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.7)

#Sorting values by descending order of highest to lowest
rules.sort_values(by=['confidence'], ascending = False)

antecedents                        consequents  \
0   (PINK REGENCY TEACUP AND SAUCER)  (GREEN REGENCY TEACUP AND SAUCER)   
1  (GREEN REGENCY TEACUP AND SAUCER)  (ROSES REGENCY TEACUP AND SAUCER)   
2  (ROSES REGENCY TEACUP AND SAUCER)  (GREEN REGENCY TEACUP AND SAUCER)   

   antecedent support  consequent support   support  confidence       lift  \
0            0.037660            0.050035  0.030910    0.820768  16.403939   
1            0.050035            0.051267  0.037553    0.750535  14.639752   
2            0.051267            0.050035  0.037553    0.732497  14.639752   

   leverage  conviction  zhangs_metric  
0  0.029026    5.300203       0.975787  
1  0.034988    3.803076       0.980765  
2  0.034988    3.551237       0.982039

In [35]:
# Collecting the inferred rules and defining a confidence level of 80%
rules = association_rules(frq_items, metric = 'confidence', min_threshold = 0.8)

#Sorting values by descending order of highest to lowest
rules.sort_values(by=['confidence'], ascending = False)

antecedents                        consequents  \
0  (PINK REGENCY TEACUP AND SAUCER)  (GREEN REGENCY TEACUP AND SAUCER)   

   antecedent support  consequent support  support  confidence       lift  \
0             0.03766            0.050035  0.03091    0.820768  16.403939   

   leverage  conviction  zhangs_metric  
0  0.029026    5.300203       0.975787

###############################################        REPORT      #########################################################
For the Basket_UK dataset, i read the csv file and preprocessed the data by dropping the invoice_No column. To extract the frequent item sets in the basket_UK dataset, i set the min_support level to 0.03 for an efficient output of 25 rows, after some adjustment from 0.15 which gave an output of 3 rows and showed no item set association rule. min_support of 0.13 gave an output of 9 rows and showed only 2 rows of association rule which was not also efficient. Consequently, all item sets with a support level below 0.1 were not showcased. To collect the inferred rules, with the association rule code, i set the measuring metric to confidence and selected three (3) confidence levels of 0.6, 0.7 and 0.8 simultaneously. Here, all the 3 confidence levels selected produced the same first 3 output and number of rules with strong association and support.

The first most important 2 items association rule result with all the 3 confidence levels shows that if customers buy PINK REGENCY TEACUP AND SAUCER, they are 82% highly likely to buy GREEN REGENCY TEACUP AND SAUCER.

The second most important 2 items association rule result with all the 3 confidence levels shows that if customers buy GREEN REGENCY TEACUP AND SAUCER, they are 75% highly likely to buy ROSES REGENCY TEACUP AND SAUCER

The Third most important 2 items association rule shows that if customers buy ROSES REGENCY TEACUP AND SAUCER, they are also 73% highly to buy GREEN TEACUP AND SAUCER. (250 words)